src/avisi/crux/datastore.clj

```diff
@@ -1,6 +1,233 @@
+                  ^"[Lcom.google.cloud.datastore.StructuredQuery$OrderBy;" (make-array StructuredQuery$OrderBy 0))
+                ^KeyQuery$Builder (.setLimit ^Integer (int 1))
+                (.build))]
+    (if-let [latest-key (first (iterator-seq (.run datastore query)))]
+      (inc (datastore-id->tx-id latest-key))
+      1)))
+
+(defn ^Entity tx->entity [transaction-log v]
+  (let [tx-id (calc-tx-id transaction-log)]
+    (->
+      (Entity/newBuilder
+        ^Key
+        (->>
+          tx-id
+          (tx-id->datastore-id transaction-log)))
+      (.set "events" (clj->blob v))
+      (.set
+        "created"
+        (->
+          (TimestampValue/newBuilder (Timestamp/now))
+          (.setExcludeFromIndexes true)
+          (.build)))
+      (.build))))
+
+(def max-tries 10)
+
+(defn ^Entity save-tx!
+  [{:keys [^Datastore datastore]
+    :as transaction-log}
+   v]
+  (loop [tries 0]
+    (let [[type value]
+            (try [:entity (.add datastore (tx->entity transaction-log v))] (catch DatastoreException e [:exception e]))]
+      (case type
+        :entity value
+        :exception
+          (if-not
+            (and (= Code/ALREADY_EXISTS (Code/forNumber (.getCode ^DatastoreException value))) (<= tries max-tries))
+            (throw value)
+            (do (log/warn "Failed saving tx on try" tries) (recur (inc tries))))))))
+
+(def ^Integer tx-log-batch-size (int 200))
+
+(defn query-seqs [^Datastore datastore ^EntityQuery$Builder q ^Cursor after]
+  (lazy-seq
+    (let [query ^Query
+                (->
+                  (.setStartCursor q after)
+                  (.build))
+          query-results ^QueryResults (.run datastore query)]
+      (when-let [entities (doall (iterator-seq query-results))]
+        (lazy-seq (cons entities (query-seqs datastore q (.getCursorAfter query-results))))))))
+
+(defn open-tx-log!
+  [{:keys [^Datastore datastore namespace]
+    :as ds}
+   after-tx-id]
+  (let [q-builder
+          ^EntityQuery$Builder
+          (cond->
+            (->
+              ^EntityQuery$Builder (Query/newEntityQueryBuilder)
+              ^EntityQuery$Builder (.setNamespace ^String namespace)
+              ^EntityQuery$Builder (.setKind transaction-kind)
+              ^EntityQuery$Builder (.setLimit tx-log-batch-size)
+              ^EntityQuery$Builder
+              (.setOrderBy
+                (StructuredQuery$OrderBy/asc "__key__")
+                ^"[Lcom.google.cloud.datastore.StructuredQuery$OrderBy;" (make-array StructuredQuery$OrderBy 0)))
+            after-tx-id (.setFilter (StructuredQuery$PropertyFilter/gt "__key__" (tx-id->datastore-id ds after-tx-id))))
+        query-results ^QueryResults (.run datastore (.build ^EntityQuery$Builder q-builder))
+        ;; get-cursor after needs the iterator to be fully realized
+        entities (doall (iterator-seq query-results))
+        cursor ^Cursor (.getCursorAfter query-results)]
+    (cio/->cursor
+      (fn [])
```

Comment:
Empty fn?

Reply:
Dit is een callback om een bepaalde resource te sluiten maar dat hebben wij niet met datastore.